sgkit/tests/test_hwe.py

```diff
@@ -0,0 +1,160 @@
+from pathlib import Path
+from typing import Tuple
+
+import numpy as np
+import pandas as pd
+import pytest
+import xarray as xr
+from pandas import DataFrame
+from xarray import DataArray, Dataset
+
+from sgkit.stats.hwe import hardy_weinberg_p_value as hwep
+from sgkit.stats.hwe import hardy_weinberg_p_value_jit as hwep_jit
+from sgkit.stats.hwe import hardy_weinberg_p_value_vec as hwep_vec
+from sgkit.stats.hwe import hardy_weinberg_p_value_vec_jit as hwep_vec_jit
+from sgkit.stats.hwe import hardy_weinberg_test as hwep_test
+from sgkit.testing import simulate_genotype_call_dataset
+
+
+def simulate_genotype_calls(
+    n_variant: int, n_sample: int, p: Tuple[float, float, float], seed: int = 0
+) -> DataArray:
+    """Get dataset with diploid calls simulated from provided genotype distribution
+
+    Parameters
+    ----------
+    n_variant : int
+        Number of variants
+    n_sample : int
+        Number of samples
+    p : Tuple[float, float, float]
+        Genotype distribution as float in [0, 1] with order
+        homozygous ref, heterozygous, homozygous alt
+
+    Returns
+    -------
+    call_genotype : (variants, samples, ploidy) DataArray
+        Genotype call matrix as 3D array with ploidy = 2.
+    """
+    rs = np.random.RandomState(seed)
+    # Draw genotype codes with provided distribution
+    gt = np.stack(
+        [
+            rs.choice([0, 1, 2], size=n_sample, replace=True, p=p)
+            for i in range(n_variant)
+        ]
+    )
+    # Expand 3rd dimension with calls matching genotypes
+    gt = np.stack([np.where(gt == 0, 0, 1), np.where(gt == 2, 1, 0)], axis=-1)
+    return xr.DataArray(gt, dims=("variants", "samples", "ploidy"))
+
+
+def get_simulation_data(datadir: Path) -> DataFrame:
+    return pd.read_csv(datadir / "sim_01.csv")
+
+
+def test_hwep__reference_impl_comparison(datadir):
+    df = get_simulation_data(datadir)
+    cts = df[["n_het", "n_hom_1", "n_hom_2"]].values
+    p_expected = df["p"].values
+    p_actual = hwep_vec(*cts.T)
+    np.testing.assert_allclose(p_expected, p_actual)
+    p_actual = hwep_vec_jit(*cts.T)
+    np.testing.assert_allclose(p_expected, p_actual)
+
+
+def test_hwep__raise_on_negative():
+    args = [[-1, 0, 0], [0, -1, 0], [0, 0, -1]]
+    for arg in args:
```

Comment:
Nit: might be slighter neater to parameterise the test?

Reply:
https://github.com/pystatgen/sgkit/pull/76/commits/466859e584c2dbe716f92d417c87c150ffa1d278#diff-8778a759770b6a59711bea6328271b0bR66